Fix missing init of libevent threads functions

zvpunry opened the pull request:
To call libevent from multiple threads, it must be initialized by
calling evthread_use_pthreads() (or evthread_use_windows_threads()) or
libevent won't protect its internal data structures against data races.

Gearman didn't call these functions and uses multiple threads by default
which caused problems like unhandled input on sockets.

sni commented:
great, maybe that solves all kind of issues we had when running gearmand without `--threads=0`.

zvpunry commented:
> great, maybe that solves all kind of issues we had when running gearmand without `--threads=0`.

--trheads=0 causes gearmand to look up the amount of available CPUs and uses that value as thread-count (with a limit of 4 threads). See https://github.com/gearman/gearmand/blob/daab15fe0fc8da1eb2653c039ff8a1ff2df8692c/gearmand/gearmand.cc#L373

sni commented:
ok, forget what i wrote. No idea then what solved the issues :-)

zvpunry commented:
> ok, forget what i wrote. No idea then what solved the issues :-)

Maybe you used "--threads=1"?

If your problems were clients connections which became stuck after sending a request, causing timeouts. The same with "gearadmin --status" or "telnet server 4730" and typing "status" or "version". And after a restart the problem was gone, for some time. Then it was probably the same bug.

esabol commented:
Well, from what I can tell, `evthread_use_pthreads()` is only available in libevent2.

I think that means the solution is not so simple?

1. `configure.ac` needs to be modified to check for libevent2 and a `HAVE_LIBEVENT2` macro added to gear_config.h where appropriate.
2. The Makefile needs to add `-levent_pthreads` after `-levent -lpthread` if we have libevent2.
3. We need something like:
```C
#ifdef HAVE_LIBEVENT2
# include <event2/event.h>
# include <event2/thread.h>
#endif
```
near the top of `gearman/gearmand.cc`, and
4. I'd go with the following instead of the propose patch:
```C
#ifdef HAVE_LIBEVENT2
# ifdef EVTHREAD_USE_PTHREADS_IMPLEMENTED
  evthread_use_pthreads();
# elif defined(EVTHREAD_USE_WINDOWS_THREADS_IMPLEMENTED)
  evthread_use_windows_threads();
# endif
#endif
```

esabol commented:
Adding `AX_LIBEVENT2(,[AC_MSG_ERROR([Unable to find libevent2])])` to `configure.ac` right after the `AX_LIBEVENT` line takes care of (1).

For (2), I can hack it by editing `libgearman-server/libgearman-server.la` and adding '-levent_pthreads` there, but we need to automate that. I'm not clear how to do that. Anyone?

Would it be a good idea to `AC_CHECK_LIB([event_pthreads],...)` somewhere?

zvpunry commented:
Thanks. You are absolutely correct, by adding the #ifdef EVTHREAD_USE_* I broke my own fix. :-(

I'll rework that patch a bit and then I'll open a new pull request.

esabol commented:
> I'll rework that patch a bit and then I'll open a new pull request.

Do you still plan to do so, @zvpunry ?

Note: This PR was intended to address issue #262.

zvpunry commented:
I planned to do so, but after looking at autoconf/automake my motivation went somewhere else.

libevent (version < 2) is no longer used, even very old versions of debian (wheezy) with gearman 0.33 used libevent2. This makes me believe that it isn't a problem to remove support for older versions of libevent, just depend on a current libevent2. That is what you did by adding "AX_LIBEVENT2(,[AC_MSG_ERROR([Unable to find libevent2])])" to configure.ac.

The missing -levent_pthreads can be added somehow to that m4/ax_libevent.m4 and libgearman-server/include.am (or just into the include.am).

gearmand/gearmand.cc just calls gearmand_create() from libgearman-server/gearmand.cc and runs it with some user provided parameters. It seems that, at some time, it was intended to make the job server something a user could include in his own program. I don't know if this ever happened. In any case, the correct position to initialize libevent2 would be in gearmand_create() or gearmand_run() if (gearmand->threads > 0).

I didn't find any hint that gearmand runs on windows with windows threads but a lot of calls to pthread_* functions. It won't run (or even compile) on any system without pthreads. So it wouldn't make it worse if the evthread_use_windows_threads() function is ignored. Just call the evthread_use_pthreads() function and if it isn't there the build should fail.

I have these changes in a "work in progress" branch: https://github.com/zvpunry/gearmand/tree/wip 

Please have a look at it.

esabol commented:
> I planned to do so, but after looking at autoconf/automake my motivation went somewhere else.

Please clarify. Does that mean you will not be submitting a PR for this?

> libevent (version < 2) is no longer used, even very old versions of debian (wheezy) with gearman 0.33 used libevent2. This makes me believe that it isn't a problem to remove support for older versions of libevent, just depend on a current libevent2.

That seems reasonable. I don't see `/usr/include/event2/` on CentOS 6, but maybe the `libevent-devel` package just needs to be installed there.

> I didn't find any hint that gearmand runs on windows with windows threads but a lot of calls to pthread_* functions. It won't run (or even compile) on any system without pthreads. So it wouldn't make it worse if the evthread_use_windows_threads() function is ignored.

Yes, I came to the same conclusion about Windows support and agree.

> I have these changes in a "work in progress" branch: https://github.com/zvpunry/gearmand/tree/wip
> 
> Please have a look at it.

This looks pretty good to me. The only glaring thing I noticed is that, unless you have evidence to the contrary, `-levent_pthreads` should come after `-levent -lpthread` in `include.am`, and you have it coming before `@PTHREAD_LIBS@`.

zvpunry commented:
> Please clarify. Does that mean you will not be submitting a PR for this?
I'll create another pull request. But don't like to touch the autoconf/automake stuff (except those two lines I already touched in my work in progress branch and one of these lines is copied from your first message).

> > libevent (version < 2) is no longer used, even very old versions of debian (wheezy) with gearman 0.33 used libevent2. This makes me believe that it isn't a problem to remove support for older versions of libevent, just depend on a current libevent2.
> 
> That seems reasonable. I don't see `/usr/include/event2/` on CentOS 6, but maybe the `libevent-devel` package just needs to be installed there.

I wouldn't expect the header files to be present if libevent-devel isn't installed. But I don't know CentOS except that it is more ancient then Debian old-stable. It's likely that they still have libevent 1.

...
> > I have these changes in a "work in progress" branch: https://github.com/zvpunry/gearmand/tree/wip
> > Please have a look at it.
> 
> This looks pretty good to me. The only glaring thing I noticed is that, unless you have evidence to the contrary, `-levent_pthreads` should come after `-levent -lpthread` in `include.am`, and you have it coming before `@PTHREAD_LIBS@`.

I'll change this.

zvpunry commented:
> > I have these changes in a "work in progress" branch: https://github.com/zvpunry/gearmand/tree/wip
> > Please have a look at it.
> 
> This looks pretty good to me. The only glaring thing I noticed is that, unless you have evidence to the contrary, `-levent_pthreads` should come after `-levent -lpthread` in `include.am`, and you have it coming before `@PTHREAD_LIBS@`.

I have updated my wip branch (with commit --amend and push --force). I moved the ordering of the linked libraries, please have another look.

esabol commented:
The libtest/has.cc change should be a separate issue, right?

Instead of explicitly specifying `-levent_pthreads` in `include.am`, I would add a definition for `LIBEVENT2_THREADSLIB` in `m4/ax_libevent.m4` and then reference it as `@LIBEVENT2_THREADSLIB@` in `include.am`. That seems straightforward, no?

Maybe add `#include <event2/thread.h>` to the test program run in `AX_LIBEVENT2`?

Do we really want the '#error' in `gearmand.cc`? I would print a warning to the logfile and set the number of threads to 1. More work, I know....

zvpunry commented:
I have updated my work in progress branch https://github.com/zvpunry/gearmand/tree/wip

Please have a look at it.

I have also a https://github.com/zvpunry/gearmand/tree/libevent2-initialization-fix and a https://github.com/zvpunry/gearmand/tree/libtest-strncpy-fix branch. They are just the result of "git cherry-pick".

Should I open an issue for the strncpy fix?

zvpunry commented:
I have pushed a new version of  https://github.com/zvpunry/gearmand/tree/libevent2-initialization-fix

Please have a look. :)

I used your warning message if thread support isn't available. I haven't changed the library order (it was correct). But I have cleaned up some mistakes I made in m4/ax_libevent.m4 and I also made some coding-style changes.

esabol commented:
> Should I open an issue for the strncpy fix?

No, that seems unnecessary.

zvpunry commented:
@esabol Did you have a look at https://github.com/zvpunry/gearmand/commit/45bd95402d71c583c4737496cc0b48930d548c14

esabol commented:
I hadn’t, but I just did. I’m by no means an expert at m4/autoconf stuff, but it looks ok to me. Go ahead and submit a PR. Thanks for seeing this through!

zvpunry commented:
OK, I have submitted the new pull request: https://github.com/gearman/gearmand/pull/268
